design-proposals/eim-decouple-microos-storage.md

```diff
@@ -0,0 +1,185 @@
+This should significantly improve the EN provisioning time.
+
+## Proposal
+
+### Challenge
+
+At a high-level, this design proposal moves the MicroOS image storage from the orchestrator to an external storage, such as Release File Server (e.g., S3 bucket)
+or a local storage (e.g., a local, on-prem file server). The orchestrator will no longer be responsible for storing the MicroOS image.
+Instead, it will only provide a URL to the image stored in an external storage.
+
+Currently, the HookOS image is curated by DKAM and stored on the PV. The curation includes the following operations:
+
+1. Downloads the pre-built HookOS image from the Release Service 
+2. Prepares the HookOS configuration file, that contains a set of orchestrator-specific configuration parameters (e.g., proxy settings, URLs, extra hosts), 
+3. Generates secure keys and signs all the components of the image (vmlinuz, initramfs)
+4. Compresses the HookOS image into the .tar.gz archive 
+5. Reads orchestrator CA certificates and injects into the Micro-OS image during build
+6. Injects environment variables into the Micro-OS image during build
+7. Re-signs the Micro-OS image after CA certificates and environment variables were injected
+
+The solution proposed in this document assumes that the Micro-OS image is stored **outside** of the orchestrator. It means that
+DKAM should not do any of the steps above. Therefore, the main challenge to solve is how to provide CA certs and environment variables to the Micro-OS image downloaded from an external source,
+without sacrificing the security.
+
+### Solution
+
+The solution assumes the following:
+
+- The MicroOS image is stored in an external storage (e.g., S3 bucket) and the orchestrator provides a URL to the image. The Micro-OS image can be stored on any other external or local server.
+  It could be a Release Service (S3 bucket) or any other HTTPS file server.
+- The Micro-OS image should be signed and the signing keys (DER keys) should be stored along with the Micro-OS image.
+- The image stored externally is a generic image, without any orchestrator-specific files embedded.
+- The orchestrator (DKAM) will have to provide orchestrator-specific files to ENs. The files package contains orchestrator CA certificate and
+  environment files that are needed for Micro-OS to work with the given Edge Orchestrator instance. The files package will be called a "Micro-OS bundle" hereinafter.
+- The Micro-OS bundle is a signed archive. It is signed by the same keychain as iPXE script.
+- iPXE script along with the Micro-OS bundle are still stored on the orchestrator PV. Since these are small files, the provisioning time would not be impacted.
+- ENs will download the Micro-OS bundle and the Micro-OS image during the iPXE phase and inject the Micro-OS bundle to the Micro-OS image (via `initrd`), similarly to how the initramfs is loaded.
```

Comment:
IHMO this is not going to work. Did someone do a TR??

Reply:
What is this cert used for?

Reply:
This is orchestrator CA certificate that is needed for Micro-OS services (device discovery, tink worker, ...) to connect to orchestrator